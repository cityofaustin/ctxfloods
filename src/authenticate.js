const authenticate = function(lokka, userType, email, password) {
  describe('When signing in as a ' + userType, () => {
    describe('with the correct username and password', () => {
      it('should generate a valid jwtToken', async () => {
        const response = await lokka.send(`
          mutation($email:String!, $password:String!) {
            authenticate(input: {email: $email, password: $password}) {
              jwtToken
            }
          }
        `,
        {
          email: email,
          password: password
        });

        expect(response.authenticate.jwtToken).not.toBeNull();
      });
    });

    describe('with an incorrect username and password', () => {
      it('should generate a null jwtToken', async () => {
        const response = await lokka.send(`
          mutation($email:String!, $password:String!) {
            authenticate(input: {email: $email, password: $password}) {
              jwtToken
            }
          }
        `,
        {
          email: email,
          password: "wrong"
        });

        expect(response.authenticate.jwtToken).toBeNull();
      });
    });
  });
}

export default authenticate;
