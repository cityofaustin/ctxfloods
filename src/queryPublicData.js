// TODO: Figure out how to use this in the community editor/admin/super admin tests
const queryPublicData = function(lokka) {
  describe('When querying public data', () => {

    it('should get all users', async () => {
      const response = await lokka.send(`
      {
        allUsers {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all communities', async () => {
      const response = await lokka.send(`
      {
        allCommunities {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all statuses', async () => {
      const response = await lokka.send(`
      {
        allStatuses {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all status updates', async () => {
      const response = await lokka.send(`
      {
        allStatusUpdates {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all status reasons', async () => {
      const response = await lokka.send(`
      {
        allStatusReasons {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all status durations', async () => {
      const response = await lokka.send(`
      {
        allStatusDurations {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all crossings', async () => {
      const response = await lokka.send(`
      {
        allCrossings {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get all community crossings', async () => {
      const response = await lokka.send(`
      {
        allCommunityCrossings {
          nodes {
            id
          }
        }
      }
      `);

      expect(response).not.toBeNull();
    });

    it('should get the latest statuses correctly', async () => {
      const response = await lokka.send(`
        {
          allCrossings {
            nodes {
              latestStatus {
                id
              }
            }
          }
        }
      `);

      expect(response).not.toBeNull();
    });

  });
}

export default queryPublicData;
