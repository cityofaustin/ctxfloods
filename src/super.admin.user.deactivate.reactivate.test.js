import HttpTransport from 'lokka-transport-http';
import Lokka from 'lokka';
import uuidv4 from 'uuid';
import queryPublicData from './queryPublicData';
import authenticate from './authenticate';

const anonLokka = new Lokka({transport: new HttpTransport('http://localhost:5000/graphql')});
const superAdminEmail = 'superadmin@flo.ods';
const superAdminPassword = 'texasfloods';
const newUserPassword = 'texasfloods';

async function getToken(email, password) {
  const response = await anonLokka.send(`
    mutation($email:String!, $password:String!) {
      authenticate(input: {email: $email, password: $password}) {
        jwtToken
      }
    }
  `,
  {
    email: email,
    password: password
  });

  return response.authenticate.jwtToken;
}

describe('When registering, deactivating, and reactivating a user as a super admin', () => {
  var newUserEmail = uuidv4() + '@flo.ods';
  var newUserId;

  describe('As a super admin', async () => {
    var lokka;

    beforeAll(async (done) => {
      getToken(superAdminEmail, superAdminPassword).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport('http://localhost:5000/graphql', {headers})});
        done();
      });
    });

    it('should register a new user', async () => {
      const response = await lokka.send(`
        mutation($email:String!) {
          registerUser(input: {
            firstName: "New",
            lastName: "User",
            jobTitle: "Community Editor",
            communityId: 1,
            phoneNumber: "555-6666",
            email: $email,
            password:"texasfloods",
            role:"floods_community_editor"
          }) {
            user {
              id
              active
            }
          }
        }
      `,
      {
        email: newUserEmail
      });

      expect(response).not.toBeNull();
      expect(response.registerUser.user.active).toBeTruthy();
    });

  });

  describe('As the new user', async () => {
    var lokka;

    beforeEach(async (done) => {
      getToken(newUserEmail, newUserPassword).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport('http://localhost:5000/graphql', {headers})});
        done();
      });
    });

    it('should get the correct current user', async () => {
      const response = await lokka.send(`
        {
          currentUser {
            id
            firstName
          }
        }
      `);

      newUserId = response.currentUser.id;
      expect(response.currentUser.firstName).toMatchSnapshot();
    });
  });

  describe('As a super admin again', async () => {
    var lokka;

    beforeAll(async (done) => {
      getToken(superAdminEmail, superAdminPassword).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport('http://localhost:5000/graphql', {headers})});
        done();
      });
    });

    it('should deactivate the user', async () => {
      const response = await lokka.send(`
        mutation($userID:Int!) {
          deactivateUser(input: {userId: $userID}) {
            user {
              id
            }
          }
        }
      `,
      {
        userID: newUserId
      });

      expect(response).not.toBeNull();
    });

  });

  describe('As the new user again', async () => {
    var lokka;

    beforeEach(async (done) => {
      getToken(newUserEmail, newUserPassword).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport('http://localhost:5000/graphql', {headers})});
        done();
      });
    });

    it('should fail to get the correct current user', async () => {
      try {
        const response = await lokka.send(`
          {
            currentUser {
              id
              emailAddress
            }
          }
        `);
      } catch(e) {
        expect(e).toMatchSnapshot();
      }
    });
  });

  it('should see the user is deactivated', async () => {
    const response = await anonLokka.send(`
      query($userId:Int!){
        userById(id:$userId) {
          firstName
          lastName
          active
        }
      }
    `,
    {
      userId: newUserId
    });

    expect(response).toMatchSnapshot();
  });

  describe('As a super admin once more', async () => {
    var lokka;

    beforeAll(async (done) => {
      getToken(superAdminEmail, superAdminPassword).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport('http://localhost:5000/graphql', {headers})});
        done();
      });
    });

    it('should reactivate the user', async () => {
      const response = await lokka.send(`
        mutation($userId:Int!, $email:String!) {
          reactivateUser(input: {
            userId: $userId,
            email: $email
            password: "texasfloods"
            role: "floods_community_editor"
          }) {
            user {
              firstName
              lastName
              active
            }
          }
        }
      `,
      {
        userId: newUserId,
        email: newUserEmail
      });

      expect(response).toMatchSnapshot();
    });

  });

  describe('As the reactivated user', async () => {
    var lokka;

    beforeAll(async (done) => {
      getToken(newUserEmail, newUserPassword).then((token) => {
        const headers = {
          'Authorization': 'Bearer '+ token
        };
        lokka = new Lokka({transport: new HttpTransport('http://localhost:5000/graphql', {headers})});
        done();
      });
    });

    it('should get the correct current user', async () => {
      const response = await lokka.send(`
        {
          currentUser {
            firstName
            lastName
          }
        }
      `);

      expect(response).toMatchSnapshot();
    });
  });
});
