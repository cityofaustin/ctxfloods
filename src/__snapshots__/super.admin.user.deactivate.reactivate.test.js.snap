// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When registering, deactivating, and reactivating a user as a super admin As a super admin once more should reactivate the user 1`] = `
Object {
  "reactivateUser": Object {
    "user": Object {
      "active": true,
      "firstName": "New",
      "lastName": "User",
    },
  },
}
`;

exports[`When registering, deactivating, and reactivating a user as a super admin As the new user again should fail to get the correct current user 1`] = `[Error: Invalid status code: 403]`;

exports[`When registering, deactivating, and reactivating a user as a super admin As the new user should get the correct current user 1`] = `"New"`;

exports[`When registering, deactivating, and reactivating a user as a super admin As the reactivated user should get the correct current user 1`] = `
Object {
  "currentUser": Object {
    "firstName": "New",
    "lastName": "User",
  },
}
`;

exports[`When registering, deactivating, and reactivating a user as a super admin should see the user is deactivated 1`] = `
Object {
  "userById": Object {
    "active": false,
    "firstName": "New",
    "lastName": "User",
  },
}
`;
